src/data_loader.py

```diff
@@ -0,0 +1,160 @@
+        all_masks = np.zeros((768, 768), dtype = np.int16)
+        #if isinstance(in_mask_list, list):
+        for mask in in_mask_list:
+            if isinstance(mask, str):
+                all_masks += self.rle_decode(mask)
+        return np.expand_dims(all_masks, -1)
+    
+    
+    def retrive_image_matrix(self, image_id):
+        # google platform (TODO!!!!!!!!!)
+        image_path = path.join(path.dirname(__file__), '../data/train_v2/', image_id)
+
+        # kaggle image path
+        #image_path = "/kaggle/input/airbus-ship-detection/train_v2/" + image_id
+        img = PIL.Image.open(image_path)
+        image_mat = np.array(img)
+        return image_mat
+        
+    def __getitem__(self, idx):
+        image_id = self.images_list[idx]
+        no_ship_list = image_dict["w/ships"]
+        image_mat = self.retrive_image_matrix(image_id)
+        # dataloader for classification
+        if self.loader_type == "classification":
+            if image_id in no_ship_list:
+                sample = {"image": image_mat, "class": 0}
+            else:
+                sample = {"image": image_mat, "class": 1}
+        
+        # dataloader for segmentation 
+        else: 
+            mask_list = image_dict["ships"][image_id]
+            mask = self.masks_as_image(mask_list) 
+            sample = {"image": image_mat, "mask": mask}
+
+        
+        # remove image id if sample without replacement
+        if self.replacement:
+            self.images_list.remove(image_id)
+
+        if self.transform:
+            sample = self.transform(sample)
+        
+        return sample
+    
+    def __random_sampler__(self):
+        n_image_list = len(self.images_list)
+        
+        # randomly choose a idx
+        random_idx = random.randint(0,n_image_list)
+        sample = self.__getitem__(random_idx)
+        return sample
+
+
+
+# helper functions
+
+# input: file name of the segmentation csv file
+# output: a dictionary {"ships": {"imageId": list of segmentation pixels}， “w/ships”: ["imageId"]}
+def build_images_dict(csv_fn):
+    """build dictionaries of images with/without ships"""
+    
+    data = pd.read_csv(csv_fn)
+    no_ships_list = []
+    ship_dict = collections.defaultdict(list)  # {"imageId": "segmentation pixels"}
+    for index, row in data.iterrows():
+        image_id = row["ImageId"]
+        encoded_pixels = row["EncodedPixels"]
+        if pd.isnull(data.loc[index, "EncodedPixels"]) :
+            no_ships_list.append(image_id)
+        else:
+            ship_dict[image_id].append(encoded_pixels)
+    image_dict = dict({"ships": ship_dict, "w/ships": no_ships_list})
+    return(image_dict)
+
+
+# preprocess the image dict to have same images with/without ships
+def process_images_dict(image_dict):
+    n = len(image_dict["ships"])
+    image_dict["w/ships"] = random.choices(image_dict["w/ships"], k=n)
+    return image_dict
+
+
```

Comment:
The code below needs to be removed. Otherwise, it's run every time we try to import the file.
If we want to add some testing, see an example [here](https://github.com/grvk/airbus-challenge/blob/dev/george/trainer/tests/tests_trainer.py). Since we're short on time though, as long as we test it manually, automated tests can be skipped